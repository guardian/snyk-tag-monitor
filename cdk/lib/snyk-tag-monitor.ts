import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import type {GuScheduledLambdaProps} from "@guardian/cdk/lib/patterns"
import {GuScheduledLambda} from "@guardian/cdk/lib/patterns"
import type { App} from "aws-cdk-lib";
import { Duration } from "aws-cdk-lib";
import { Schedule } from "aws-cdk-lib/aws-events";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import {Secret} from "aws-cdk-lib/aws-secretsmanager";
import { Topic } from "aws-cdk-lib/aws-sns";
import { EmailSubscription } from "aws-cdk-lib/aws-sns-subscriptions";


export class SnykTagMonitor extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const app = "snyk-tag-monitor"

    const topic = new Topic(this, `${app}-topic`)
    topic.addSubscription(new EmailSubscription("devx.security@guardian.co.uk"))
    const lambdaProps: GuScheduledLambdaProps = {
      rules: [{ schedule: Schedule.rate(Duration.days(2))}],
      monitoringConfiguration: {
        toleratedErrorPercentage: 50,
        snsTopicName: topic.topicName,
      },
      runtime: Runtime.PYTHON_3_9,
      handler: "main.hello",
      app,
      fileName: `${app}.zip` 
    }

    const snykTagLambda = new GuScheduledLambda(this,
      app, lambdaProps)

    const snykApiKey = new Secret(this, "snykApiKey")
    const snykGroupId = new Secret(this, "snykGroupId")

  }
}
