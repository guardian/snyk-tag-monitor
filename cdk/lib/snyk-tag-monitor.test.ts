import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { SnykTagMonitor } from "./snyk-tag-monitor";

describe("The SnykTagMonitor stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new SnykTagMonitor(app, "SnykTagMonitor", { stack: "snyk-tag-monitor", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
