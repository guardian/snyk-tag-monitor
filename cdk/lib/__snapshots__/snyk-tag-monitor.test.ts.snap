// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SnykTagMonitor stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuAlarm",
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuLambdaErrorPercentageAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "snyktagmonitor01C2294D": {
      "DependsOn": [
        "snyktagmonitorServiceRoleDefaultPolicy2C0C2161",
        "snyktagmonitorServiceRole8EBA8B3B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "snyk-tag-monitor/TEST/snyk-tag-monitor/snyk-tag-monitor.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "snyk-tag-monitor",
            "SNS_TOPIC_ARN": {
              "Ref": "snyktagmonitortopicF2FA58D7",
            },
            "STACK": "snyk-tag-monitor",
            "STAGE": "TEST",
          },
        },
        "Handler": "main.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "snyktagmonitorServiceRole8EBA8B3B",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Tags": [
          {
            "Key": "App",
            "Value": "snyk-tag-monitor",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/snyk-tag-monitor",
          },
          {
            "Key": "Stack",
            "Value": "snyk-tag-monitor",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "snyktagmonitorErrorPercentageAlarmForLambda2893A757": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "snyktagmonitortopicF2FA58D7",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "snyktagmonitor01C2294D",
              },
              " exceeded 50% error rate",
            ],
          ],
        },
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "High error % from ",
              {
                "Ref": "snyktagmonitor01C2294D",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "snyktagmonitor01C2294D",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "snyktagmonitor01C2294D",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "snyktagmonitor01C2294D",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 50,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "snyktagmonitorEventInvokeConfig5F4D7B41": {
      "Properties": {
        "FunctionName": {
          "Ref": "snyktagmonitor01C2294D",
        },
        "MaximumRetryAttempts": 1,
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "snyktagmonitorServiceRole8EBA8B3B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "snyk-tag-monitor",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/snyk-tag-monitor",
          },
          {
            "Key": "Stack",
            "Value": "snyk-tag-monitor",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "snyktagmonitorServiceRoleDefaultPolicy2C0C2161": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/snyk-tag-monitor/TEST/snyk-tag-monitor/snyk-tag-monitor.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/snyk-tag-monitor/snyk-tag-monitor",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/snyk-tag-monitor/snyk-tag-monitor/*",
                  ],
                ],
              },
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "snyktagmonitortopicF2FA58D7",
              },
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "snyktagmonitorServiceRoleDefaultPolicy2C0C2161",
        "Roles": [
          {
            "Ref": "snyktagmonitorServiceRole8EBA8B3B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "snyktagmonitoralarm178C7D05": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "snyktagmonitortopicF2FA58D7",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "stage",
            "Value": "TEST",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "snykTagCount",
        "Namespace": "snyk-tag-monitor",
        "Period": 86400,
        "Statistic": "Minimum",
        "Threshold": 4500,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "snyktagmonitorsnyktagmonitorrate1day0AllowEventRuleSnykTagMonitorsnyktagmonitorAE4097CFB7932521": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "snyktagmonitor01C2294D",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "snyktagmonitorsnyktagmonitorrate1day0D20FBF83",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "snyktagmonitorsnyktagmonitorrate1day0D20FBF83": {
      "Properties": {
        "ScheduleExpression": "rate(1 day)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "snyktagmonitor01C2294D",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "snyktagmonitortopicF2FA58D7": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/snyk-tag-monitor",
          },
          {
            "Key": "Stack",
            "Value": "snyk-tag-monitor",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "snyktagmonitortopicdevxsecurityguardiancoukCC41C9E5": {
      "Properties": {
        "Endpoint": "devx.security@guardian.co.uk",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "snyktagmonitortopicF2FA58D7",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
}
`;
